### PR TITLE
add type-safe tag_t to represent fix tags

### PR DESCRIPTION
**Pull request by arvidn**
to prevent tags to be mixed with raw ints.
Also annotate the tag_t with the argument type the tag expects, to improve type safety of the push_back functions.

How controversial would a patch along these lines be?
it's really two somewhat independent changes.
1. introduce fix_t, to remove access to int operations that are nonsensical on tags.
2. introduce a template type-tag to disable certain push_back functions for certain tags. i.e. don't allow pushing a string to an integer tag and vice versa. The types could be made a bit more granular by distinguishing between integer and float types.

### CODE REVIEW

**Comment by arvidn** [comment]
I seem to recall you saying you're targeting C++11. Do I misremember that?

**Comment by jamesdbrock** [comment]
First, this is some really nice template C++ that you're written, and thank you very much for the contribution. Second, I'm totally in general agreement with you about the importance of type-safety, in fact, these days I'm more into Haskell than C++.

However, my choice of `int` for the type of tags was a decision I made deliberately. https://github.com/jamesdbrock/hffix#user-defined-fields-and-custom-tags

> High Frequency FIX Parser does not enforce the data type of the Field Definitions for content fields in the FIX spec, so the developer is free to read or write any tag number with any field data type.

The FIX protocol specification is a poorly-conceived mess and implementations that exist in the wild can have surprising proprietary "extensions." It's my feeling that type constraint of the tag types of the `message_writer::push_back` family of functions will be more likely to prevent a legitimate intentional tag assignment than to catch an important mistake at compile-time.

Plus, reading and writing any field as a string value should be allowed by the hffix library, because sometimes the users may want to do their own string-casting, instead of relying of the hffix library string-casting functions. For example, `writer.push_back_string(tag::SendingTime, "20140928-07:12:06.000");` should be allowed. And I think `writer.push_back_int(hffix::tag::OrdType, 2);` should also be allowed.

Honestly I think that where hffix library is really weak is in reading values from integer fields. `field_value::as_int()` and `field_value::as_decimal()` should have a way to notify when the field is not a well-formed numeric string. Currently it's sort of expected that if the library user is worried that the field might not be a well-formed numeric string, then they can scan the string themselves before calling `as_int()`. There should be a better story there. Or at least better docs.

**Comment by jamesdbrock** [comment]
> I seem to recall you saying you're targeting C++11.

Well... I'm trying to target all versions from C++98 to C++17.

**Comment by arvidn** [comment]
ok. I agree about the field types, that must be flexible. I still think there's value in distinguishing ``fix_t`` from ``int`` (but without the template tag). A rule of thumb I like is that a function signature should be understandable without parameter names (but just the type names).

I will strip the patch down to what I would think provides more clear type-safety benefits, to see what you think.

File: include/hffix.hpp [hunk 1]
```diff
@@ -1968,19 +1969,20 @@ inline void message_reader_const_iterator::increment()
 /* @cond EXCLUDE */
 
 namespace details {
-inline bool is_tag_a_data_length(int tag)
+inline bool is_tag_a_data_length(tag_t const tag)
 {
-    int* length_fields_end = length_fields + (sizeof(length_fields)/sizeof(length_fields[0]));
-    return std::find(length_fields, length_fields_end, tag) != length_fields_end; // fields are ordered, so this could be std::binary_search.
+    // fields are ordered, so this could be std::binary_search.
+    return std::find(std::begin(length_fields), std::end(length_fields), tag) != std::end(length_fields);
```
Comment (arvidn) [inline include/hffix.hpp]: oops, one more c++11 feature

**Comment by jamesdbrock** [comment]
Yeah I agree in principle that

> A rule of thumb I like is that a function signature should be understandable without parameter names (but just the type names)

The `class tag_t` with explicit constructor and casting operator will deliberately prevent use case _1_ of these three:

1. `writer.push_back_string (35, "A");`
2. `writer.push_back_string (tag_t(35), "A");`
3. `writer.push_back_string (tag::MsgType, "A");`

I personally do not hate case _1_ and don't want to statically forbid people from ever doing it.

What about replacing `class tag_t` with this?

```
typedef int tag_t
```
This allows all three use cases.

**Comment by arvidn** [comment]
using a typedef would defeat the purpose, at least the purpose in my mind, which is:

1. prevent operator+, operator-, operator/,operator* etc. on tag values, since it is extremely likely to be a mistake.
2. prevent passing a variable of type int as the tag, even though it holds a value of some other quantity/unit.

The signature taking a typedef is identical as one taking an ``int``.

In modern C++ (i.e. C++11) case 1 would be handled by:

``writer.push_back_string(35_tag, "a");``

(on a side note though, I also think all the push back functions should all be called ``push_back``, to support using them in generic code).

**Comment by jamesdbrock** [comment]
> In modern C++ (i.e. C++11) case 1 would be handled by

Oh yeah, cool, I forgot about user-defined numeric literals in C++11.

> prevent operator+, operator-, operator/,operator* etc. on tag values, since it is extremely likely to be a mistake.

I can think of at least one case where one might want to do addition to tags. The *data* field type is actually represented by a pair of fields, where the first field is a string rendering of the byte length of the second field. And it's always true, though never specified, that the *data field tag = data length field tag + 1*. (See notes for http://jamesdbrock.github.io/hffix/classhffix_1_1message__writer.html#a1af8e5acf8ded56677fd4b6dca0bdb10 )

I can't personally think of a good reason to check that relationship at runtime, but I don't want to statically prevent everyone else from doing that kind of arithmetic. And considering how weird FIX protocol is, I wouldn't be surprised if there are other situations in which one might want to reason arithmetically about tags. I think tags are best modeled as an additive group.

You may be right that arithmetic on tags is "extremely likely to be a mistake." It's tough to say. My instinct is that it's more likely to be deliberate. I've seen too many situations where the type system is used to overconstrain. I would much rather err on the side of allowing everything which might possibly be legit. Sorry, I would really like merge this idea since you've obviously put some time and thought into it, but I have kind of strong feelings about it. I also like how just using `int` is vastly simpler than the alternatives.

> I also think all the push back functions should all be called `push_back`

Yeah I agree. And I tried to write the library that way originally, but I found it's not easy to make the type signature distinguish the function. For example, `push_back_int` and `push_back_char`. Or `push_back_decimal` and `push_back_monthyear`. So I ended up with this design with overloaded families of `push_back` functions.

**Pull request closed without merging**